### PR TITLE
Fix argument default value (Ruby Date) not being serialized

### PR DESCRIPTION
**Pull request by TomasBarry**
This commit addresses an issue whereby using a Ruby Date object as the
default value for an argument was being serialized to `nil` instead of
either serializing to a `Date` or raising an error. The fix contained in
this commit serializes the object as a `Date`.

Fixes https://github.com/rmosolgo/graphql-ruby/issues/3520

### CODE REVIEW

**Comment by TomasBarry** [comment]
@rmosolgo - are you happy with this PR? If there are any changes/missing specs I'll add them or if there's any more context required let me know.

**Comment by andrei-byte** [comment]
yes I am happy

File: lib/graphql/types/iso_8601_date.rb [hunk 1]
```diff
@@ -21,13 +21,22 @@ def self.coerce_result(value, _ctx)
...
+          value.to_date
+        else
+          Date.iso8601(value)
+        end
       rescue ArgumentError, TypeError
-        # Invalid input
-        nil
+        # binding.pry
+        err = GraphQL::DateEncodingError.new(value)
+        ctx.schema.type_error(err, ctx)
```
[the remark this answers is not in the document]
Comment (TomasBarry) [reply lib/graphql/types/iso_8601_date.rb:38]: @rmosolgo:

1. Is the default behaviour not guaranteed to be maintained based on existing tests? Or is there a specific example you want me to test for?
2. Which definition of `type_error` are you looking for me to test exactly? I can do but not sure which one you want me to test/what you want tested as I can't find an existing test.

File: lib/graphql/types/iso_8601_date.rb [hunk 2]
```diff
@@ -21,13 +21,22 @@ def self.coerce_result(value, _ctx)
         Date.parse(value.to_s).iso8601
       end
 
-      # @param str_value [String]
+      # @param str_value [String | Date | DateTime | Time]
```
[the remark this answers is not in the document]
Comment (TomasBarry) [reply lib/graphql/types/iso_8601_date.rb]: Sorted.

File: lib/graphql/types/iso_8601_date.rb [hunk 3]
```diff
@@ -21,13 +21,22 @@ def self.coerce_result(value, _ctx)
...
+          value.to_date
+        elsif value.is_a?(::Time)
+          value.to_date
+        else
+          Date.iso8601(value)
+        end
       rescue ArgumentError, TypeError
-        # Invalid input
-        nil
+        # binding.pry
```
[the remark this answers is not in the document]
Comment (TomasBarry) [reply lib/graphql/types/iso_8601_date.rb]: Sorted.

File: lib/graphql/types/iso_8601_date.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (TomasBarry) [reply lib/graphql/types/iso_8601_date.rb:38]: @rmosolgo - looking further, the default behaviour should be the same considering [this](https://github.com/rmosolgo/graphql-ruby/pull/3563/files#diff-bb7e26f3df072a57fe4393fde6e1e8c2b5c03c302664d94e233d35c81fa462c4R62) spec. Unless I'm missing something?

[the remark this answers is not in the document]
Comment (TomasBarry) [reply lib/graphql/types/iso_8601_date.rb:38]: Same goes for [this](https://github.com/rmosolgo/graphql-ruby/pull/3563/files#diff-bb7e26f3df072a57fe4393fde6e1e8c2b5c03c302664d94e233d35c81fa462c4R62) spec.

**Comment by rmosolgo** [comment]
Thanks again for your work on this! I updated the branch and added another test to double-check that `.type_error` handling was working as expected.

**Pull request merged**